File: back/src/main/java/com/back/domain/ticket/entity/Ticket.java
package com.back.domain.ticket.entity;

import com.back.domain.schedule.entity.Schedule;
import com.back.domain.schedule.entity.ScheduleSeat;
import com.back.domain.user.entity.User;
import com.back.global.jpa.entity.BaseEntity;
import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;

@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Ticket extends BaseEntity {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long ticketId;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "schedule_id", nullable = false)
    private Schedule schedule;

    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "schedule_seat_id", nullable = false)
    private ScheduleSeat scheduleSeat;

    @Column(nullable = false, unique = true)
    private String ticketNumber;

    @Column(nullable = false)
    private int ticketPrice;

    @Column(nullable = false)
    private boolean isValid = true;

   private Ticket(User user,Schedule schedule,ScheduleSeat scheduleSeat,String ticketNumber,int ticketPrice,boolean isValid) {
       this.user = user;
       this.schedule = schedule;
       this.scheduleSeat = scheduleSeat;
       this.ticketNumber = ticketNumber;
       this.ticketPrice = ticketPrice;
       this.isValid = isValid;
   }
   public static Ticket create(User user,Schedule schedule,ScheduleSeat scheduleSeat,String ticketNumber,int ticketPrice) {
       return new Ticket(user,schedule,scheduleSeat,ticketNumber,ticketPrice,true);
   }
    public void updateIsValid(boolean isValid) {
        this.isValid = isValid;
    }
}

File: back/src/main/kotlin/com/back/domain/ticket/entity/Ticket.kt
package com.back.domain.ticket.entity

import com.back.domain.schedule.entity.Schedule
import com.back.domain.schedule.entity.ScheduleSeat
import com.back.domain.user.entity.User
import com.back.global.exception.ErrorCode
import com.back.global.exception.ServiceException
import com.back.global.jpa.entity.BaseEntity
import jakarta.persistence.*

@Entity
class Ticket(
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    val user: User,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "schedule_id", nullable = false)
    val schedule: Schedule,

    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "schedule_seat_id", nullable = false)
    val scheduleSeat: ScheduleSeat,

    @Column(nullable = false, unique = true)
    val ticketNumber: String,

    @Column
    val groupToken: String? = null,

    @Column(nullable = false)
    val ticketPrice: Int = 0,

    isValid: Boolean = true
) : BaseEntity() {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val ticketId: Long? = null

    @Column(nullable = false)
    var isValid: Boolean = isValid
        protected set

    fun cancel() {
        if (!isValid) {
            throw ServiceException(ErrorCode.TICKET_ALREADY_CANCELLED)
        }
        this.isValid = false
    }

    companion object {
        fun create(
            user: User,
            schedule: Schedule,
            scheduleSeat: ScheduleSeat,
            ticketNumber: String,
            ticketPrice: Int,
            groupToken: String
        ): Ticket = Ticket(
            user = user,
            schedule = schedule,
            scheduleSeat = scheduleSeat,
            ticketNumber = ticketNumber,
            groupToken = groupToken,
            ticketPrice = ticketPrice,
            isValid = true
        )
    }
}
